walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,53 +15,90 @@
-  // space complexity is O(max(E, G))
+  // space complexity is O(E)
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
     // similar to merge intervals problem. My idea is to merge all request's attendees' events
     // and consider gaps between those events, if the gap is big enough to fit request's time, add the gap to the answer
-    ArrayList<TimeRange> answer = new ArrayList<>();
-    ArrayList<String> attendees = new ArrayList<>(request.getAttendees());
+    // Follow up on optional attendees
+    // I guess the algorithm is the following:
+    // merge answer with the optional events, and if new answer is not empty, return new answer
+    // otherwise ignore optional attendees
     ArrayList<Event> relevantEvents = new ArrayList<>();
+    // the list of events of optional employees'
+    ArrayList<Event> optionalEvents = new ArrayList<>();
 
     // check if request for a valid duration
     if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
-      return answer;
+      return Arrays.asList();
     }
-  
-    for (String p: attendees) {
-      // get the list of all events for this person
-      for (Event e: events) {
+
+    // get the list of events for required attendees
+    for (String p : request.getAttendees()) {
+      for (Event e : events) {
         // if event includes the person
         if (e.getAttendees().contains(p)) {
           relevantEvents.add(e);
         }
       }
     }
 
+    // get the list of events for optional attendees
+    for (String p : request.getOptionalAttendees()) {
+      for (Event e : events) {
+        // if event includes the person
+        if (e.getAttendees().contains(p)) {
+          optionalEvents.add(e);
+        }
+      }
+    }
+
     // sort events by start date
     relevantEvents.sort((e1, e2) -> {
       TimeRange e1when = e1.getWhen();
       TimeRange e2when = e2.getWhen();
       if (e1when.start() == e2when.start()) {
-        return e1when.end() - e2when.start();
+        return e1when.end() - e2when.end();
       }
       return e1when.start() - e2when.start();
-    });
-    // if no events, return whole day
-    if (relevantEvents.size() == 0) {
-      answer.add(TimeRange.WHOLE_DAY);
+    });    
+    // now merge all events
+    ArrayList<Interval> takenIntervals = getTakenIntervals(relevantEvents);
+    ArrayList<TimeRange> answer = getFreeTimeRanges(takenIntervals, request.getDuration());
+    ArrayList<TimeRange> answerIncludingOptionals = answer;
+    // handle optionals by inserting each of their events to the taken intervals, and if
+    // there are no free gaps remains, return noOptionalAnswer
+    for (Event e: optionalEvents) {
+      // check if some free space left that can accomodate the request
+      takenIntervals = insertInterval(takenIntervals, eventToInterval(e));
+      answerIncludingOptionals = getFreeTimeRanges(takenIntervals, request.getDuration());
```

Comment:
It should... but I think we can do it the other way also - we can check here `answerIncludingOptionals` size and break the loop earlier in case the size is 0. I mean there is no point to continue inserting intervals if there are no free time ranges already

Reply:
Sounds good too, please update.